substate-store-impls/src/hash_tree/mod.rs

```diff
@@ -28,8 +28,59 @@ mod test;
 #[allow(dead_code)]
 mod types;
 
+/// Payload stored at each leaf node.
+#[derive(Clone, Debug, Ord, PartialOrd, Hash, Eq, PartialEq)]
+pub struct ValuePayload {
+    /// State version at which the stored value hash was most recently changed.
+    ///
+    /// In practice, this is a state version at which:
+    /// - (in Substate Tier) the represented Substate was most recently upserted,
+    /// - or (in upper Tiers) there was any change in the represented lower Tier.
+    ///
+    /// Please note this can be different than the tree Node's version (i.e. there are cases in which a tree Node is
+    /// changed - e.g. gets promoted up after losing all siblings - while its value remains unchanged).
+    ///
+    /// The version recorded here is used to navigate the 3-Tier JMT structure (i.e. locate the right lower tier root).
+    pub last_hash_change_version: Version,
+
+    /// The actual value of the represented Substate.
+    ///
+    /// May be empty:
+    /// - for leaf nodes that do not represent Substates (i.e. leafs of upper Tiers),
+    /// - or when the JMT instance is explicitly configured to not store values.
+    ///
+    /// TODO(potential refactoring): This field is relevant only for Substate Tier, so from object modelling PoV, it
```

Comment:
Actually I think it would probably be a generic `T: TreeStructure` with associated types `VersionedNode`, `Payload` and `PayloadUpdate`

Reply:
I do agree that going "full generic" would make the code more explicit / type-safe. With some care, we could even elegantly use the same generic infrastructure to drive the "store only hashes / store substate values" behavior.

I am still hesitating whether this would make things more readable, though.

> a generic T: TypedTree with associated types VersionedNode, Payload and PayloadUpdate and methods to combine them.

This ☝️ sentence does not promise immediate simplifications :(
_(I mean: the road to it may be long, and the result may still be disappointing)_

Meanwhile, a sentence "A leaf has an optional `DbSubstateValue` field, which is populated on Substate tier if you set this flag." is conceptually very easy, but results in this ugly code that we see here :(

Which is why I am still torn.

Either way, it is totally a topic for a separate PR.